### ISSUE TITLE
Some useful tree constructors

### ISSUE THREAD

**Issue by liufengyun** [lampepfl/dotty#5567]
In implementing ScalaTest macros, I find two tree constructors are useful:

```Scala
def selectApply(receiver: Term, name: Name, args: List[Term]): Term
def let(expr: Term)(body: Ident => Term): Term
```

The constructor `selectApply` should do _overloading resolution_ internally, but no other adapation (like inserting `apply` nor implicits). The constructor `let` allows creating bindings friendly.

The current implementation of `assert` is as follows:

```Scala
      case Term.Apply(Term.Select(lhs, op), rhs :: Nil) =>
        op match {
          case "==" =>
            val left = lhs.seal[Any]
            val right = rhs.seal[Any]
            '{
              val _left   = ~left
              val _right  = ~right
              val _result = _left == _right
              val _bool = Bool.binaryMacroBool(_left, ~op.toExpr, _right, _result, ~prettifier)
              Assertions.assertionsHelper.macroAssert(_bool, ~clue, ~pos)
            }
          case ">" =>
            // blocked by tasty constructors
            // https://github.com/lampepfl/dotty/pull/5438
            val left = lhs.seal[Int]
            val right = rhs.seal[Int]
            '{
              val _left   = ~left
              val _right  = ~right
              val _result = _left > _right
              val _bool = Bool.binaryMacroBool(_left, ~op.toExpr, _right, _result, ~prettifier)
              Assertions.assertionsHelper.macroAssert(_bool, ~clue, ~pos)
            }
```

With the two constructors, the implementation can be made more generic & simpler:

```Scala
case Term.Apply(Term.Select(lhs, op), rhs :: Nil) =>
  val res =
  let(lhs) { left =>
    let(rhs) { right =>
      let(selectApply(left, op, right)) { result =>
         val l = left.seal[Any]
         val r = right.seal[Any]
         val b = result.seal[Boolean]
        '{
            val _bool = Bool.binaryMacroBool(~l, ~op.toExpr, ~r, ~b, ~prettifier)
            Assertions.assertionsHelper.macroAssert(_bool, ~clue, ~pos)      
         }
    }
  }
  res.seal[Assertion]
```

**Comment by nicolasstucki** [comment]
Those should not be implemented in the core. They should be implemented in some library.  The let abstraction is already in https://github.com/nicolasstucki/quoted-util

**Comment by smarter** [comment]
> Those should not be implemented in the core. They should be implemented in some library. 

Why ?

**Comment by liufengyun** [comment]
@nicolasstucki Just checked https://github.com/nicolasstucki/quoted-util, it seems they are not tasty reflect util, but `quote` util. I assume implementing overloading resolution will be a challenge in library?

**Comment by nicolasstucki** [comment]
Oh, yes. Those are not on trees. Though they are safer.

**Comment by nicolasstucki** [comment]
Here is the precise link https://github.com/nicolasstucki/quoted-util/blob/master/src/main/scala/quoted/util/Let.scala#L22-L25

**Comment by liufengyun** [comment]
The provided constructors are not enough:

- We lack API to create let-bindings (val-bindings) dynamically
- The way to support dynamic select on a term based on `copy` is awkward
- The semantics of `Select.copy` is unspecified in the case of overloading

We cannot use quote to introduce the binding, because we do not know the type of `lhs` and `rhs` statically.

```Scala
        case Term.Apply(sel @ Term.Select(lhs, op), rhs :: Nil) =>
          case ">" =>
            val Term.IsSelectModule(select) = sel
            def app(l: Expr[Any], r: Expr[Any]): Expr[Boolean] =
              Term.apply(Term.Select.copy(select)(Term.Apply(fn, l.unseal :: Nil), ">"), r.unseal :: Nil).seal[Boolean]

            val left = lhs.seal[???]
            val right = rhs.seal[???]
            '{
              val _left   = ~left
              val _right  = ~right
              val _result = ~app('_left, '_right)
              Bool.binaryMacroBool(_left, ~op.toExpr, _right, _result, ~prettifier)
            }
```

It will be really nice to have something like `let` and `selectApply` or equivalent as proposed in the issue.

**Comment by nicolasstucki** [comment]
#5651 shows how to implement `def let(expr: Term)(body: Ident => Term): Term` in library code.

**Comment by liufengyun** [comment]
> Those should not be implemented in the core. They should be implemented in some library. 

I think core and non-core are just a problem of separation, there are multiple ways to achieve the separation, e.g. different packages, not neccesarily in different libraries.
